> # <b>Practical Lab 1 - Univariate Linear Regression <b>

#### **Prediction for California Housing Prices Dataset Using Linear regression**

> **Student Name**: Hasyashri Bhatt

>**Student ID**: 9028501

>**Course**: Foundations of Machine Learning Frameworks(CSCN8010) 

>**GitHub Repository**: https://github.com/Hasyashri/MLFramework_CSCN8010/tree/main/Assignment



> ## Introduction

In this Lab I will train first Machine Learning Model. There are multiple algorithms to train the model but here I will apply the `univariate linear regression`, which is a simple yet powerful algorithm that models the `relationship` between one `**independent variable**` and a `**dependent variable**` using a `linear equation: Y = mX + c`. <span style="font-size:10px">Refered from :https://github.com/CSCN8010/CSCN8010/tree/main/class_notebooks/linear_regression</span>

I will explore how the median house value varies with three different factors. The independent variables considered in this study are:
| Model | Independent Variable   | Description                  |
|-------|------------------------|------------------------------|
| 1     | Median Income          | Average income in the Dataset   |
| 2     | Population             | Total population  |
| 3     | Number of Households   | Count of households in the Dataset|


> # Step 1 : Framing the Problem
**Objective**
The objective of this lab is to develop and compare three univariate linear regression models to predict median house value. Each model uses a different independent variable: median income, population, and number of households. 

**Goal**
Through this analysis, I aim to identify which variable is the strongest predictor of housing prices and gain insights about it.I will train three separate models, each using a different feature as below:

1. Mean Squared Error (MSE)
2. Mean Absolute Error (MAE)
3. R² Score

> # Step 2 : Getting the Data

I have downloaded dataset from the below link and loaded in the Pandas.
**California Housing Prices Dataset Link**:https://www.kaggle.com/datasets/camnugent/california-housing-prices 

In [44]:
# Importing necessary libraries
import pandas as pd             # loading and manipulating data
import numpy as np              # numerical operations
import matplotlib.pyplot as plt # plotting
import seaborn as sns           # data visualization


**Load the dataset into the Panda Dataframe here I will initialize Dataframe as a df.**

In [45]:
# Load the dataset from a CSV file to a pandas DataFrame
df = pd.read_csv(r'C:\Users\habha\OneDrive\Desktop\Assignment Linear\Hasyashri-MLFramework_CSCN8010\Assignment\housing.csv')
# If it's loaded correctly, print successfully loaded message
if df is not None:
    print("Data loaded successfully.")

Data loaded successfully.


In [46]:
# Show the first few rows of the DataFrame
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [47]:
df.info()  # Display information about the DataFrame, including data types and non-null counts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


**Check if there are any missing Values**

In [52]:
df.isnull().sum()  # Check for missing values in each column

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [49]:
df.shape  # Get the shape of the DataFrame (rows, columns)

(20640, 10)

**After Checking the Dataset I get to know that there are missing values in the total_bedroom column.I need fill this missing values in the Dataset. For that I am using fillna() method.**

In [54]:
# Fill missing values in 'total_bedrooms' column with the median value of that column
df['total_bedrooms'] = df['total_bedrooms'].fillna(df['total_bedrooms'].median())


In [56]:
# Verify that there are no more missing values in 'total_bedrooms'
print("Missing values in 'total_bedrooms':", df['total_bedrooms'].isnull().sum())

Missing values in 'total_bedrooms': 0


Before there are missing values:`207` present in the `total_bedrooms` column, now after using median of the column I filled the missing values in the `total_bedroom`column using fillna() method which resulted in: `Missing values in 'total_bedrooms': 0`.

Now going to check if there are any duplicated rows are present above we checked total rows and columns of the dataset using `.shape` which is (20640, 10).

In [61]:
# check for duplicates in the DataFrame
df.duplicated().sum()  # Count the number of duplicate rows

0

There are no duplicate rows or data present in the Dataset as it shows 0 as result.

In [58]:
# Importing machine learning libraries later when its needed
from sklearn.linear_model import LinearRegression # linear regression model
from sklearn.model_selection import train_test_split # splitting data into training and testing sets
from sklearn.metrics import mean_squared_error # model evaluation
from IPython.display import display, Markdown # display in Jupyter notebooks